File: src/main/java/nl/knaw/huygens/alexandria/lmnl/modifier/LMNLModifier.java
package nl.knaw.huygens.alexandria.lmnl.modifier;

import java.util.Collection;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.knaw.huygens.alexandria.lmnl.data_model.Limen;
import nl.knaw.huygens.alexandria.lmnl.data_model.TextNode;
import nl.knaw.huygens.alexandria.lmnl.data_model.TextRange;

public class LMNLModifier {

  Logger LOG = LoggerFactory.getLogger(LMNLModifier.class);

  private Limen limen;

  public LMNLModifier(Limen limen) {
    this.limen = limen;
  }

  public void addTextRange(TextRange newTextRange, Position position) {
    // logTextNodes(limen.textNodeList);
    // logTextRanges(limen.textRangeList);
    TextNodeCursor cursor = new TextNodeCursor(limen);
    int startOffset = position.getOffset();
    int endOffset = position.getOffset() + position.getLength();
    // get to starting TextNode
    // logTextNodes(limen.textNodeList);
    boolean findEndingTextNode = handleStartingTextNode(newTextRange, cursor, startOffset);
    // logTextRanges(limen.textRangeList);
    // logTextNodes(limen.textNodeList);
    if (findEndingTextNode) {
      handleEndingTextNode(newTextRange, cursor, endOffset);
    }
    // logTextNodes(limen.textNodeList);
    // limen.textRangeList.add(newTextRange);
    // logTextRanges(limen.textRangeList);
  }

  private boolean handleStartingTextNode(TextRange newTextRange, TextNodeCursor cursor, int startOffset) {
    boolean findStartingTextNode = true;
    boolean findEndingTextNode = true;
    while (findStartingTextNode) {
      String currentText = cursor.getCurrentText();
      int currentTextNodeLength = cursor.getCurrentTextLength();
      int offsetAtEndOfCurrentTextNode = cursor.getOffsetAtEndOfCurrentTextNode();
      if (startOffset < offsetAtEndOfCurrentTextNode) {
        // newTextRange starts in this TextNode
        int tailLength = Math.min(offsetAtEndOfCurrentTextNode, offsetAtEndOfCurrentTextNode - startOffset);
        int headLength = currentTextNodeLength - tailLength;
        if (headLength == 0) {
          // newTextRange exactly covers current TextNode
          newTextRange.addTextNode(cursor.getCurrentTextNode());
          limen.associateTextWithRange(cursor.getCurrentTextNode(), newTextRange);
          findEndingTextNode = false;
        } else {
          if (tailLength > 0) {
            // detach tail
            String headText = currentText.substring(0, headLength);
            String tailText = currentText.substring(headLength);
            cursor.getCurrentTextNode().setContent(headText);
            TextNode newTailNode = new TextNode(tailText);
            TextNode nextTextNode = cursor.getCurrentTextNode().getNextTextNode();
            newTailNode.setPreviousTextNode(cursor.getCurrentTextNode());
            newTailNode.setNextTextNode(nextTextNode);
            if (nextTextNode != null) {
              nextTextNode.setPreviousTextNode(newTailNode);
            }
            cursor.getCurrentTextNode().setNextTextNode(newTailNode);
            limen.getTextRanges(cursor.getCurrentTextNode()).forEach(tr -> {
              tr.addTextNode(newTailNode);
              limen.associateTextWithRange(newTailNode, tr);
            });
            newTextRange.addTextNode(newTailNode);
            limen.associateTextWithRange(newTailNode, newTextRange);
            limen.textNodeList.add(cursor.getTextNodeIndex() + 1, newTailNode);

          } else {
            // newTextRange.addTextNode(cursor.getCurrentTextNode());
            // limen.associateTextWithRange(cursor.getCurrentTextNode(), newTextRange);
            throw new RuntimeException("tail=empty!");
          }
        }
        findStartingTextNode = false;

      } else {
        findStartingTextNode = cursor.canAdvance();
      }
      cursor.advance();
    }
    return findEndingTextNode;
  }

  private void handleEndingTextNode(TextRange newTextRange, TextNodeCursor cursor, int endOffset) {
    boolean findEndingTextNode = true;
    while (findEndingTextNode) {
      int offsetAtEndOfCurrentTextNode = cursor.getOffsetAtEndOfCurrentTextNode();
      if (offsetAtEndOfCurrentTextNode < endOffset) {
        // this is not the TextNode where newTextRange ends, but it is part of newTextRange
        limen.associateTextWithRange(cursor.getCurrentTextNode(), newTextRange);
        findEndingTextNode = cursor.canAdvance();
        cursor.advance();

      } else {
        // this is the TextNode where newTextRange ends
        int tailLength = offsetAtEndOfCurrentTextNode - endOffset;
        int headLength = cursor.getCurrentTextLength() - tailLength;

        if (tailLength > 0) {
          if (headLength > 0) {
            // detach tail
            String headText = cursor.getCurrentText().substring(0, headLength);
            String tailText = cursor.getCurrentText().substring(headLength);
            cursor.getCurrentTextNode().setContent(headText);
            TextNode newTailNode = new TextNode(tailText);
            TextNode nextTextNode = cursor.getCurrentTextNode().getNextTextNode();
            newTailNode.setNextTextNode(nextTextNode);
            newTailNode.setPreviousTextNode(cursor.getCurrentTextNode());
            cursor.getCurrentTextNode().setNextTextNode(newTailNode);
            limen.getTextRanges(cursor.getCurrentTextNode())//
                .stream()//
                .filter(tr -> !newTextRange.equals(tr))//
                .forEach(tr -> {
                  limen.associateTextWithRange(newTailNode, tr);
                  tr.addTextNode(newTailNode);
                });
            limen.textNodeList.add(cursor.getTextNodeIndex() + 1, newTailNode);

          } else {
            // limen.associateTextWithRange(cursor.getCurrentTextNode(), newTextRange);
            // newTextRange.addTextNode(cursor.getCurrentTextNode());
            throw new RuntimeException("head=empty!");
          }
        }
        findEndingTextNode = false;
      }
    }
  }

  public void addTextRange(TextRange newTextRange, Collection<Position> positions) {
    if (!newTextRange.hasId()) {
      throw new RuntimeException("TextRange " + newTextRange.getTag() + " should have an id.");
    }
    positions.forEach(

        position -> {
          LOG.info("position={}", position);
          logTextNodes(limen.textNodeList);
          logTextRanges(limen.textRangeList);
          addTextRange(newTextRange, position);
        });

    logTextNodes(limen.textNodeList);
    logTextRanges(limen.textRangeList);
    // LMNLImporter.joinDiscontinuedRanges(limen);
  }

  private void logTextNodes(List<TextNode> list) {
    StringBuilder textnodes = new StringBuilder();
    list.forEach(

        tn -> {
          if (tn.getPreviousTextNode() != null) {
            textnodes.append("\"").append(tn.getPreviousTextNode().getContent()).append("\" -> ");
          }
          textnodes.append("[").append(tn.getContent()).append("]");
          if (tn.getNextTextNode() != null) {
            textnodes.append(" -> \"").append(tn.getNextTextNode().getContent()).append("\"");
          }
          textnodes.append("\n");
        });
    LOG.info("\nTextNodes:\n{}", textnodes);

  }

  private void logTextRanges(List<TextRange> list) {
    StringBuilder textranges = new StringBuilder();
    list.forEach(

        tr -> {
          textranges.append("[").append(tr.getTag()).append("}\n");
          tr.textNodes.forEach(tn -> textranges.append("  \"").append(tn.getContent()).append("\"\n"));
        });
    LOG.info("\nTextRanges:\n{}", textranges);
  }

  class TextNodeCursor {
    private TextNode currentTextNode;
    private int textNodeIndex = 0;
    private int offset = 0;

    public TextNodeCursor(Limen limen) {
      currentTextNode = limen.textNodeList.get(0);
    }

    public void advance() {
      offset += getCurrentTextLength();
      currentTextNode = currentTextNode.getNextTextNode();
      textNodeIndex++;
    }

    public TextNode getCurrentTextNode() {
      return currentTextNode;
    }

    public String getCurrentText() {
      return currentTextNode.getContent();
    }

    public int getCurrentTextLength() {
      return getCurrentText().length();
    }

    public int getTextNodeIndex() {
      return textNodeIndex;
    }

    public boolean canAdvance() {
      return currentTextNode.getNextTextNode() != null;
    }

    public int getOffset() {
      return offset;
    }

    private int getOffsetAtEndOfCurrentTextNode() {
      return offset + getCurrentTextLength();
    }
  }

}

File: main/src/main/kotlin/nl/knaw/huygens/alexandria/lmnl/modifier/LMNLModifier.kt
package nl.knaw.huygens.alexandria.lmnl.modifier

/*
 * #%L
 * alexandria-markup-core
 * =======
 * Copyright (C) 2016 - 2021 HuC DI (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import nl.knaw.huygens.alexandria.data_model.Limen
import nl.knaw.huygens.alexandria.data_model.Markup
import nl.knaw.huygens.alexandria.data_model.TextNode
import org.slf4j.LoggerFactory
import kotlin.math.min

internal class LMNLModifier(private val limen: Limen) {
    private val log = LoggerFactory.getLogger(LMNLModifier::class.java)

    fun addMarkup(newMarkup: Markup, position: Position) {
        // logTextNodes(limen.textNodeList);
        // logMarkups(limen.markupList);
        val cursor = TextNodeCursor(limen)
        val startOffset = position.offset
        val endOffset = position.offset + position.length
        // get to starting TextNode
        // logTextNodes(limen.textNodeList);
        val findEndingTextNode = handleStartingTextNode(newMarkup, cursor, startOffset)
        // logMarkups(limen.markupList);
        // logTextNodes(limen.textNodeList);
        if (findEndingTextNode) {
            handleEndingTextNode(newMarkup, cursor, endOffset)
        }
        // logTextNodes(limen.textNodeList);
        // limen.markupList.add(newMarkup);
        // logMarkups(limen.markupList);
    }

    private fun handleStartingTextNode(newMarkup: Markup, cursor: TextNodeCursor, startOffset: Int): Boolean {
        var findStartingTextNode = true
        var findEndingTextNode = true
        while (findStartingTextNode) {
            val currentText = cursor.currentText
            val currentTextNodeLength = cursor.currentTextLength
            val offsetAtEndOfCurrentTextNode: Int = cursor.offsetAtEndOfCurrentTextNode
            if (startOffset < offsetAtEndOfCurrentTextNode) {
                // newMarkup starts in this TextNode
                val tailLength = min(offsetAtEndOfCurrentTextNode, offsetAtEndOfCurrentTextNode - startOffset)
                val headLength = currentTextNodeLength - tailLength
                if (headLength == 0) {
                    // newMarkup exactly covers current TextNode
                    newMarkup.addTextNode(cursor.currentTextNode)
                    limen.associateTextWithRange(cursor.currentTextNode, newMarkup)
                    findEndingTextNode = false
                } else {
                    if (tailLength > 0) {
                        // detach tail
                        val headText = currentText.substring(0, headLength)
                        val tailText = currentText.substring(headLength)
                        cursor.currentTextNode.content = headText
                        val newTailNode = TextNode(tailText)
                        val nextTextNode = cursor.currentTextNode.nextTextNode
                        newTailNode.previousTextNode = cursor.currentTextNode
                        newTailNode.nextTextNode = nextTextNode
                        if (nextTextNode != null) {
                            nextTextNode.previousTextNode = newTailNode
                        }
                        cursor.currentTextNode.nextTextNode = newTailNode
                        limen.getMarkups(cursor.currentTextNode).forEach { tr: Markup ->
                            tr.addTextNode(newTailNode)
                            limen.associateTextWithRange(newTailNode, tr)
                        }
                        newMarkup.addTextNode(newTailNode)
                        limen.associateTextWithRange(newTailNode, newMarkup)
                        limen.textNodeList.add(cursor.textNodeIndex + 1, newTailNode)
                    } else {
                        // newMarkup.addTextNode(cursor.getCurrentTextNode());
                        // limen.associateTextNodeWithMarkupForLayer(cursor.getCurrentTextNode(), newMarkup);
                        throw RuntimeException("tail=empty!")
                    }
                }
                findStartingTextNode = false
            } else {
                findStartingTextNode = cursor.canAdvance()
            }
            cursor.advance()
        }
        return findEndingTextNode
    }

    private fun handleEndingTextNode(newMarkup: Markup, cursor: TextNodeCursor, endOffset: Int) {
        var findEndingTextNode = true
        while (findEndingTextNode) {
            val offsetAtEndOfCurrentTextNode: Int = cursor.offsetAtEndOfCurrentTextNode
            if (offsetAtEndOfCurrentTextNode < endOffset) {
                // this is not the TextNode where newMarkup ends, but it is part of newMarkup
                limen.associateTextWithRange(cursor.currentTextNode, newMarkup)
                findEndingTextNode = cursor.canAdvance()
                cursor.advance()
            } else {
                // this is the TextNode where newMarkup ends
                val tailLength = offsetAtEndOfCurrentTextNode - endOffset
                val headLength = cursor.currentTextLength - tailLength
                if (tailLength > 0) {
                    if (headLength > 0) {
                        // detach tail
                        val headText = cursor.currentText.substring(0, headLength)
                        val tailText = cursor.currentText.substring(headLength)
                        cursor.currentTextNode.content = headText
                        val newTailNode = TextNode(tailText)
                        val nextTextNode = cursor.currentTextNode.nextTextNode
                        newTailNode.nextTextNode = nextTextNode
                        newTailNode.previousTextNode = cursor.currentTextNode
                        cursor.currentTextNode.nextTextNode = newTailNode
                        limen.getMarkups(cursor.currentTextNode)
                                .stream()
                                .filter { tr: Markup -> newMarkup != tr }
                                .forEach { tr: Markup ->
                                    limen.associateTextWithRange(newTailNode, tr)
                                    tr.addTextNode(newTailNode)
                                }
                        limen.textNodeList.add(cursor.textNodeIndex + 1, newTailNode)
                    } else {
                        // limen.associateTextNodeWithMarkupForLayer(cursor.getCurrentTextNode(), newMarkup);
                        // newMarkup.addTextNode(cursor.getCurrentTextNode());
                        throw RuntimeException("head=empty!")
                    }
                }
                findEndingTextNode = false
            }
        }
    }

    fun addMarkup(newMarkup: Markup, positions: Collection<Position>) {
        if (!newMarkup.hasId()) {
            throw RuntimeException("Markup " + newMarkup.tag + " should have an id.")
        }
        positions.forEach { position: Position ->
            log.debug("position={}", position)
            logTextNodes(limen.textNodeList)
            logMarkups(limen.markupList)
            addMarkup(newMarkup, position)
        }
        logTextNodes(limen.textNodeList)
        logMarkups(limen.markupList)
        // LMNLImporter.joinDiscontinuedRanges(limen);
    }

    private fun logTextNodes(list: List<TextNode>) {
        val textnodes = StringBuilder()
        list.forEach { tn: TextNode ->
            if (tn.previousTextNode != null) {
                textnodes.append("\"").append(tn.previousTextNode.content).append("\" -> ")
            }
            textnodes.append("[").append(tn.content).append("]")
            if (tn.nextTextNode != null) {
                textnodes.append(" -> \"").append(tn.nextTextNode.content).append("\"")
            }
            textnodes.append("\n")
        }
        log.debug("\nTextNodes:\n{}", textnodes)
    }

    private fun logMarkups(list: List<Markup>) {
        val markups = StringBuilder()
        list.forEach { tr: Markup ->
            markups.append("[").append(tr.tag).append("}\n")
            tr.textNodes.forEach { tn: TextNode -> markups.append("  \"").append(tn.content).append("\"\n") }
        }
        log.debug("\nMarkups:\n{}", markups)
    }

    internal class TextNodeCursor(limen: Limen) {
        var currentTextNode: TextNode
            private set
        var textNodeIndex = 0
            private set
        var offset = 0
            private set

        fun advance() {
            offset += currentTextLength
            currentTextNode = currentTextNode.nextTextNode
            textNodeIndex++
        }

        val currentText: String
            get() = currentTextNode.content

        val currentTextLength: Int
            get() = currentText.length

        fun canAdvance(): Boolean {
            return currentTextNode.nextTextNode != null
        }

        internal val offsetAtEndOfCurrentTextNode: Int
            get() = offset + currentTextLength

        init {
            currentTextNode = limen.textNodeList[0]
        }
    }

}
